### PR TITLE
Amend manifest search to work well with multi-projects builds

### PR DESCRIPTION
**Pull request by futtetennista**
In my multi-project build, replicant would incorrectly set the package name of one of the sub-projects (which is a library project). I amended the implementation so that it looks for the manifest in the current directory and all sub-directories and checks if the sub-project is an app project. 

I couldn't really figure out how to do a couple of things properly, so suggestions are welcome:
1. if the current project is a library project, the package will be set to the one of the sample app, which is probably not what you want. But how can replicant know that's the case? A possibility would be to look for know app directory names like "sample", skip those and fall back to the library project.
2. how to write tests for this, on the other hand there weren't any for the previous implementation.

### CODE REVIEW

File: lib/replicant/repl.rb [hunk 1]
```diff
@@ -154,18 +154,34 @@ def setup_rlwrap
...
-          return path if path.include?(manifest_file)
+          if File.directory?(path)
+            manifest_file_path = known_manifest_locations.map {|known_location| "#{path}/#{known_location}"}.find {|loc| File.exist?(loc)}
+            return manifest_file_path if manifest_file_path and is_application_path?(path)
+          end
         end
       end
     end
 
+    def is_application_path?(path)
```
Comment (mttkay) [inline lib/replicant/repl.rb]: just a style thing, but the `?` in Ruby already denotes that this is a boolean property. Drop `is_`

File: lib/replicant/repl.rb [hunk 2]
```diff
@@ -154,18 +154,34 @@ def setup_rlwrap
...
+      if File.exists?("#{path}/project.properties") # old library project
+        return IO.readlines("#{path}/project.properties").find { |l| l =~ /android.library=true/ }.nil?
```
Comment (mttkay) [inline lib/replicant/repl.rb]: this looks a bit convoluted. Can this be replaced with: `return IO.readlines(...).none? { |l| l =~ ...  }`

Comment (mttkay) [inline lib/replicant/repl.rb]: http://ruby-doc.org/core-2.1.5/Enumerable.html#method-i-none-3F

File: lib/replicant/repl.rb [hunk 3]
```diff
@@ -154,18 +154,34 @@ def setup_rlwrap
...
+      else
+        # TODO: support gradle files named something other than build.gradle.
+        if File.exists?("#{path}/build.gradle")
+          return not(IO.readlines("#{path}/build.gradle").find { |l| l =~ /com.android.application/ }.nil?)
```
Comment (mttkay) [inline lib/replicant/repl.rb]: here, too, double negation is very difficult to read... ("not something is nil")

Could you use `any?`

```
return IO.readlines(...).any? { |l| l =~ ...}  
```

Comment (mttkay) [inline lib/replicant/repl.rb]: also the else-if can be collapsed into an `elsif`

**Comment by mttkay** [comment]
Thanks! :+1:

**Comment by mttkay** [comment]
> how to write tests for this, on the other hand there weren't any for the previous implementation.

yeah I know, it's difficult to test. I used to have specs that actually test the repl output but this was very brittle. I reverted to only testing command logic, not configuration (this is mostly configuration and setup logic, which is difficult to test)

I wouldn't lose sleep over it

**Comment by futtetennista** [comment]
Thanks for the feedback!

**Comment by mttkay** [comment]
:+1:

**Pull request merged**